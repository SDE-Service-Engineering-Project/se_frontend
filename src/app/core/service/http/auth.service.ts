import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {AuthResponse} from "../../model/AuthResponse";

const AUTH_API = 'http://localhost:8080/auth/';

@Injectable({
  providedIn: 'root'
})
export class AuthService {

  constructor(private http: HttpClient) {
  }

  public login(userName: string, password: string): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(
      AUTH_API + 'login',
      {userName, password}
    );
  }
}
