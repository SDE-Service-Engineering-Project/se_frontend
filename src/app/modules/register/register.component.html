<div class="container-fluid center-screen">
  <div id="main-wrapper" class="container">
    <div class="row justify-content-center">
      <div class="col-xl-10">
        <div class="card border-0">
          <div class="card-body p-0">
            <div class="row no-gutters">
              <div class="col-lg-6">
                <div class="p-5">
                  <div class="mb-5">
                    <h3 class="h4 font-weight-bold text-theme" style="color: #5369f8">Sign Up</h3>
                  </div>
                  <h6 class="h5 mb-0">You finally made it!</h6>
                  <p class="text-muted mt-2 mb-5">Sign up for the worlds leading car rental service.</p>

                  <form [formGroup]="registerForm" id="registerForm" (ngSubmit)="submitRegister()">

                    <div class="form-group" style="text-align: left">
                      <label for="inputUserName">Username</label>
                      <input type="text" class="form-control" id="inputUserName" formControlName="username"
                             [ngClass]="{ 'is-invalid': registerForm.get('username')?.touched && registerForm.get('username')!!.errors, 'is-valid': registerForm.get('username')!!.valid }"
                      >
                      <div *ngIf="registerForm.get('username')!!.touched && registerForm.get('username')!!.errors"
                           class="invalid-feedback">
                        <div *ngIf="registerForm.get('username')!!.errors!!.hasOwnProperty('required')"
                             id="err-username">
                          Username is required!
                        </div>
                      </div>
                    </div>

                    <div class="form-group" style="text-align: left">
                      <label for="inputFirstName">Firstname</label>
                      <input type="text" class="form-control" id="inputFirstName" formControlName="firstname"
                             [ngClass]="{ 'is-invalid': registerForm.get('firstname')?.touched && registerForm.get('firstname')!!.errors, 'is-valid': registerForm.get('firstname')!!.valid }"
                      >
                      <div *ngIf="registerForm.get('firstname')!!.touched && registerForm.get('firstname')!!.errors"
                           class="invalid-feedback">
                        <div *ngIf="registerForm.get('firstname')!!.errors!!.hasOwnProperty('required')"
                             id="err-firstName">
                          Firstname is required!
                        </div>
                      </div>
                    </div>

                    <div class="form-group" style="text-align: left">
                      <label for="inputLastName">Lastname</label>
                      <input type="text" class="form-control" id="inputLastName" formControlName="lastname"
                             [ngClass]="{ 'is-invalid': registerForm.get('lastname')?.touched && registerForm.get('lastname')!!.errors, 'is-valid': registerForm.get('lastname')!!.valid }"
                      >
                      <div *ngIf="registerForm.get('lastname')!!.touched && registerForm.get('lastname')!!.errors"
                           class="invalid-feedback">
                        <div *ngIf="registerForm.get('lastname')!!.errors!!.hasOwnProperty('required')"
                             id="err-lastname">
                          Lastname is required!
                        </div>
                      </div>
                    </div>


                    <div class="form-group" style="text-align: left">
                      <label for="inputPassword">Password</label>
                      <input type="password" class="form-control" id="inputPassword" formControlName="password"
                             [ngClass]="{ 'is-invalid': registerForm.get('password')!!.touched && registerForm.get('password')!!.errors, 'is-valid': registerForm.get('password')!!.valid }"
                      >
                      <div *ngIf="registerForm.get('password')!!.touched && registerForm.get('password')!!.errors"
                           class="invalid-feedback">
                        <div *ngIf="registerForm.get('password')!!.errors!!.hasOwnProperty('required')"
                             id="err-password">
                          Password is required!
                        </div>
                      </div>
                    </div>

                    <div class="form-group mb-5" style="text-align: left">
                      <label for="inputConfirmPassword">Confirm Password</label>
                      <input type="password" class="form-control" id="inputConfirmPassword"
                             formControlName="confirm_password"
                             [ngClass]="{ 'is-invalid': registerForm.get('confirm_password')!!.touched && registerForm.get('confirm_password')!!.errors, 'is-valid': registerForm.get('confirm_password')!!.valid }"
                      >
                      <div
                        *ngIf="registerForm.get('confirm_password')!!.touched && registerForm.get('confirm_password')!!.errors"
                        class="invalid-feedback">
                        <div *ngIf="registerForm.get('confirm_password')!!.errors!!"
                             id="err-confirm_password">
                          Password does not match!
                        </div>
                      </div>
                    </div>

                    <div class="form-group mb-5">
                      <button type="submit" class="btn btn-theme form-control" data-testid="register-btn"
                              id="register-btn"
                              [disabled]="registerForm!!.invalid">Register
                      </button>
                    </div>
                  </form>
                </div>
              </div>
              <div class="col-lg-6 d-none d-lg-inline-block">
                <div class="logo-register rounded-right">
                  <div class="overlay rounded-right"></div>
                  <div class="logo-inner">
                    <h4 class="text-white mb-4">Welcome to Car Rental</h4>
                    <p class="lead text-white">"Float like a Cadillac, sting like a Beemer."</p>
                    <p>- Lightning McQueen</p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <p class="text-muted text-center mt-3 mb-0">Already have an account? Login <a
          routerLink="" rel="noopener" class="text-primary ml-1">here</a>.
        </p>
      </div>
    </div>
  </div>
</div>
