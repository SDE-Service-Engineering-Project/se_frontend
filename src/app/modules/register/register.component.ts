import { Component } from '@angular/core';
import {
  AbstractControl,
  FormControl,
  FormGroup,
  ValidationErrors,
  ValidatorFn,
  Validators,
} from '@angular/forms';
import { Router } from '@angular/router';
import { AuthService } from '../../services/http/auth.service';
import { ToastService } from '../../services/toast/toast.service';
import { HttpErrorResponse } from '@angular/common/http';

@Component({
  selector: 'app-register',
  templateUrl: './register.component.html',
  styleUrls: ['./register.component.sass'],
})
export class RegisterComponent {
  registerForm = new FormGroup({
    username: new FormControl('', [
      Validators.required,
      Validators.minLength(3),
      Validators.maxLength(32),
    ]),
    firstname: new FormControl('', [
      Validators.required,
      Validators.minLength(3),
      Validators.maxLength(30),
    ]),
    lastname: new FormControl('', [
      Validators.required,
      Validators.minLength(3),
      Validators.maxLength(30),
    ]),
    password: new FormControl('', [
      Validators.required,
      Validators.minLength(8),
      Validators.maxLength(30),
      matchValidator('confirm_password', true),
    ]),
    confirm_password: new FormControl('', [
      Validators.required,
      Validators.minLength(8),
      Validators.maxLength(30),
      matchValidator('password'),
    ]),
  });

  constructor(
    private authService: AuthService,
    private router: Router,
    private toastService: ToastService
  ) {}

  submitRegister() {
    if (this.registerForm.invalid) return;

    this.toastService.removeAll();

    this.authService
      .register(
        this.registerForm.get('username')?.value!!,
        this.registerForm.get('firstname')?.value!!,
        this.registerForm.get('lastname')?.value!!,
        this.registerForm.get('password')?.value!!
      )
      .subscribe({
        next: (test) => {
          this.router
            .navigate([''])
            .then(() =>
              this.toastService.showDefaultSuccessToast(
                'Account successfully created'
              )
            );
        },
        error: (err: HttpErrorResponse) => {
          this.registerForm.get('password')?.setValue('');
          this.registerForm.get('confirm_password')?.setValue('');
          this.toastService.showDefaultErrorToast(err.error.message);
        },
      });
  }
}

export function matchValidator(
  matchTo: string,
  reverse?: boolean
): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    if (control.parent && reverse) {
      const c = (control.parent?.controls as any)[matchTo] as AbstractControl;
      if (c) {
        c.updateValueAndValidity();
      }
      return null;
    }
    return !!control.parent &&
      !!control.parent.value &&
      control.value === (control.parent?.controls as any)[matchTo].value
      ? null
      : { matching: true };
  };
}
