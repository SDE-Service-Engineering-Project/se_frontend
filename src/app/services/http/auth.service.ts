import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {AuthResponse} from '../../models/AuthResponse';

export const AUTH_API = 'http://localhost:8080/api/v1/auth/';

@Injectable({
  providedIn: 'root',
})
export class AuthService {
  constructor(private http: HttpClient) {
  }

  public login(userName: string, password: string): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(AUTH_API + 'login', {
      userName,
      password,
    });
  }

  public register(userName: string, firstName: string, lastName: string, password: string): Observable<any> {
    return this.http.post(AUTH_API + 'signup', {
      userName,
      firstName,
      lastName,
      password,
    });
  }

  public refreshToken(refreshToken: string, userName: string): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(AUTH_API + 'refresh/token', {
      refreshToken,
      userName,
    });
  }
}
