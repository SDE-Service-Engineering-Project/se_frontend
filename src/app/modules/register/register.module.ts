import {NgModule} from "@angular/core";
import {RegisterRoutingModule} from "./config/register.routing";
import {CommonModule} from "@angular/common";
import {ReactiveFormsModule} from "@angular/forms";
import {RegisterComponent} from "./register.component";

@NgModule({
  imports: [RegisterRoutingModule, CommonModule, ReactiveFormsModule],
  declarations: [RegisterComponent]
})
export class RegisterModule {

}
