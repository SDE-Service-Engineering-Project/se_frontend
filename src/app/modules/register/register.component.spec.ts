import {ComponentFixture} from '@angular/core/testing';

import {RegisterComponent} from './register.component';
import {createComponentFactory, Spectator} from "@ngneat/spectator/jest";
import {HttpClientTestingModule} from "@angular/common/http/testing";
import {ReactiveFormsModule} from "@angular/forms";

describe('RegisterComponent', () => {
  let component: RegisterComponent;
  let fixture: ComponentFixture<RegisterComponent>;
  let spectator: Spectator<RegisterComponent>;

  const createComponent = createComponentFactory({
    component: RegisterComponent,
    imports: [HttpClientTestingModule, ReactiveFormsModule],
  });

  beforeEach(() => {
    spectator = createComponent();
    fixture = spectator.fixture;
    component = spectator.component;
  });

  it('should render the component', () => {
    expect(fixture.debugElement.nativeElement).toMatchSnapshot();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should enable register button when all fields are correctly set', function () {
    component.registerForm.get('username')?.setValue('username');
    component.registerForm.get('firstname')?.setValue('firstname');
    component.registerForm.get('lastname')?.setValue('lastname');
    component.registerForm.get('password')?.setValue('password');
    component.registerForm.get('confirm_password')?.setValue('password');

    spectator.detectChanges();

    expect(fixture.debugElement.nativeElement.querySelector('#register-btn').disabled).toBeFalsy();

  });

  // TODO: add register test with mocks and routes
});
