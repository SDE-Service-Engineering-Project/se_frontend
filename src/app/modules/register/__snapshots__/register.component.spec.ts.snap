// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterComponent should render the component 1`] = `
<div
  id="root0"
>
  <div
    class="container-fluid center-screen"
  >
    <div
      class="container"
      id="main-wrapper"
    >
      <div
        class="row justify-content-center"
      >
        <div
          class="col-xl-10"
        >
          <div
            class="card border-0"
          >
            <div
              class="card-body p-0"
            >
              <div
                class="row no-gutters"
              >
                <div
                  class="col-lg-6"
                >
                  <div
                    class="p-5"
                  >
                    <div
                      class="mb-5"
                    >
                      <h3
                        class="h4 font-weight-bold text-theme"
                        style="color: #5369f8;"
                      >
                        Sign Up
                      </h3>
                    </div>
                    <h6
                      class="h5 mb-0"
                    >
                      You finally made it!
                    </h6>
                    <p
                      class="text-muted mt-2 mb-5"
                    >
                      Sign up for the worlds leading car rental service.
                    </p>
                    <form
                      class="ng-untouched ng-pristine ng-invalid"
                      id="registerForm"
                      novalidate=""
                    >
                      <div
                        class="form-group"
                        style="text-align: left;"
                      >
                        <label
                          for="inputUserName"
                        >
                          Username
                        </label>
                        <input
                          class="form-control ng-untouched ng-pristine ng-invalid"
                          formcontrolname="username"
                          id="inputUserName"
                          type="text"
                        />
                        
                      </div>
                      <div
                        class="form-group"
                        style="text-align: left;"
                      >
                        <label
                          for="inputFirstName"
                        >
                          Firstname
                        </label>
                        <input
                          class="form-control ng-untouched ng-pristine ng-invalid"
                          formcontrolname="firstname"
                          id="inputFirstName"
                          type="text"
                        />
                        
                      </div>
                      <div
                        class="form-group"
                        style="text-align: left;"
                      >
                        <label
                          for="inputLastName"
                        >
                          Lastname
                        </label>
                        <input
                          class="form-control ng-untouched ng-pristine ng-invalid"
                          formcontrolname="lastname"
                          id="inputLastName"
                          type="text"
                        />
                        
                      </div>
                      <div
                        class="form-group"
                        style="text-align: left;"
                      >
                        <label
                          for="inputPassword"
                        >
                          Password
                        </label>
                        <input
                          class="form-control ng-untouched ng-pristine ng-invalid"
                          formcontrolname="password"
                          id="inputPassword"
                          type="password"
                        />
                        
                      </div>
                      <div
                        class="form-group mb-5"
                        style="text-align: left;"
                      >
                        <label
                          for="inputConfirmPassword"
                        >
                          Confirm Password
                        </label>
                        <input
                          class="form-control ng-untouched ng-pristine ng-invalid"
                          formcontrolname="confirm_password"
                          id="inputConfirmPassword"
                          type="password"
                        />
                        
                      </div>
                      <div
                        class="form-group mb-5"
                      >
                        <button
                          class="btn btn-theme form-control"
                          disabled=""
                          id="register-btn"
                          type="submit"
                        >
                          Register 
                        </button>
                      </div>
                    </form>
                  </div>
                </div>
                <div
                  class="col-lg-6 d-none d-lg-inline-block"
                >
                  <div
                    class="logo-register rounded-right"
                  >
                    <div
                      class="overlay rounded-right"
                    />
                    <div
                      class="logo-inner"
                    >
                      <h4
                        class="text-white mb-4"
                      >
                        Welcome to Car Rental
                      </h4>
                      <p
                        class="lead text-white"
                      >
                        "Float like a Cadillac, sting like a Beemer."
                      </p>
                      <p>
                        - Lightning McQueen
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <p
            class="text-muted text-center mt-3 mb-0"
          >
            Already have an account? Login 
            <a
              class="text-primary ml-1"
              rel="noopener"
              routerlink=""
            >
              here
            </a>
            . 
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
